mlprimitives/custom/timeseries_anomalies.py

```diff
@@ -0,0 +1,169 @@
+
+
+def z_goodness(z, errors, mean, std):
+    """Compute how good a z value is.
+
+    The original goodness formula is::
+
+                 (delta_mean/mean) + (delta_std/std)
+        ------------------------------------------------------
+        number of errors above + (number of sequences above)^2
+
+    In this case, we return this value inverted (we make it negative),
+    to later on use scipy to minimize this function (instead of maximizing,
+    which was the original paper proposal).
+    """
+
+    epsilon = mean + z * std
+
+    delta_mean, delta_std = deltas(errors, epsilon, mean, std)
+    above, consecutive = count_above(errors, epsilon)
+
+    numerator = delta_mean / mean + delta_std / std
+    denominator = above + consecutive ** 2
+
+    if denominator == 0:
+        return np.inf
+
+    return -numerator / denominator
+
+
+def find_threshold(errors, min_z=0, max_z=10):
+    """Find the ideal threshold.
+
+    The ideal threshold is the one that minimizes the z_goodness function.
+    """
+
+    mean = errors.mean()
+    std = errors.std()
+
+    candidates = list()
+    for z in range(min_z, max_z):
+        best = fmin(z_goodness, z, args=(errors, mean, std), disp=False)
+        candidates.append(best[0])
+
+    best_z = candidates[0]
+    best_goodness = np.inf
+    for z in candidates:
+        goodness = z_goodness(z, errors, mean, std)
+        if goodness < best_goodness:
+            best_z = z
+            best_goodness = goodness
+
+    return mean + best_z * std
+
+
+def find_sequences(errors, epsilon):
+    """Find sequences of values that are above epsilon.
+
+    This is done following this steps:
+
+        * create a boolean mask that indicates which value are above epsilon.
+        * shift this mask by one place, filing the empty gap with a False
+        * compare the shifted mask with the original one to see if there are changes.
+        * Consider a sequence start any point which was true and has changed
+        * Consider a sequence end any point which was false and has changed
+    """
+    above = pd.Series(errors > epsilon)
+    shift = above.shift(1).fillna(False)
+    change = above != shift
+
+    index = above.index
+    starts = index[above & change].tolist()
+    ends = index[~above & change].tolist()
+    if len(ends) == len(starts) - 1:
+        ends.append(len(above))
+
+    return list(zip(starts, ends))
+
+
+def find_anomalies(errors, index, z_range=(0, 10)):
+    """Find sequences of values that are anomalous.
+
+    We first find the ideal threshold for the set of errors that we have,
+    and then find the sequences of values that are above this threshold.
+
+    Lastly, we compute a score proportional to the maximum error in the
+    sequence, and finally return the index pairs that correspond to
+    each sequence, along with its score.
+    """
+
+    mean = errors.mean()
+    std = errors.std()
+
+    threshold = find_threshold(errors, z_range[0], z_range[1])
```

Comment:
Here we are splitting `z_range`, to pass it to `find_threshold`, where only be used to generate a range, couldn't just keep it all together and call range unpacking z_range,like `range(*z_range)` on `find_threshold`?